File: src/main/java/org/codeontology/extraction/FieldEntity.java
package org.codeontology.extraction;

import com.hp.hpl.jena.rdf.model.RDFNode;
import org.codeontology.Ontology;
import spoon.reflect.declaration.CtField;
import spoon.reflect.reference.CtFieldReference;
import spoon.reflect.reference.CtTypeReference;

import java.lang.reflect.*;
import java.util.ArrayList;
import java.util.List;

public class FieldEntity extends NamedElementEntity<CtField<?>> implements ModifiableEntity<CtField<?>>, MemberEntity<CtField<?>>, TypedElementEntity<CtField<?>> {

    public FieldEntity(CtField<?> field) {
        super(field);
    }

    public FieldEntity(CtFieldReference<?> field) {
        super(field);
    }

    @Override
    public String buildRelativeURI() {
        return getDeclaringElement().getRelativeURI() + SEPARATOR + getReference().getSimpleName();
    }

    @Override
    protected RDFNode getType() {
        return Ontology.FIELD_ENTITY;
    }

    @Override
    public void extract() {
        tagName();
        tagType();
        tagDeclaringElement();
        tagJavaType();
        tagModifiers();
        if (isDeclarationAvailable()) {
            tagComment();
            tagAnnotations();
        }
    }

    @Override
    public List<Modifier> getModifiers() {
        if (isDeclarationAvailable()) {
            return Modifier.asList(getElement().getModifiers());
        } else {
            try {
                return Modifier.asList(((CtFieldReference<?>) getReference()).getModifiers());
            } catch (Exception | Error e) {
                return new ArrayList<>();
            }
        }
    }

    public void tagModifiers() {
        new ModifiableTagger(this).tagModifiers();
    }

    @Override
    public TypeEntity<?> getJavaType() {
        TypeEntity<?> type;
        if (isDeclarationAvailable()) {
            type = getFactory().wrap(getElement().getType());
        } else {
            type = getGenericType();
            if (type == null) {
                CtTypeReference<?> typeReference = ((CtFieldReference<?>) getReference()).getType();
                type = getFactory().wrap(typeReference);
            }
        }

        type.setParent(getDeclaringElement());
        return type;
    }

    private TypeEntity<?> getGenericType() {
        TypeEntity<?> result = null;
        if (!isDeclarationAvailable()) {
            try {
                CtFieldReference<?> reference = ((CtFieldReference<?>) getReference());
                Field field = (Field) reference.getActualField();
                Type genericType = field.getGenericType();

                if (genericType instanceof GenericArrayType ||
                        genericType instanceof TypeVariable<?>  ||
                        genericType instanceof ParameterizedType) {

                    result = getFactory().wrap(genericType);
                }

            } catch (Throwable t) {
                return null;
            }
        }

        return result;
    }

    public void tagJavaType() {
        new JavaTypeTagger(this).tagJavaType();
    }

    @Override
    public Entity<?> getDeclaringElement() {
        if (isDeclarationAvailable()) {
            return getFactory().wrap(getElement().getDeclaringType());
        } else {
            CtFieldReference<?> reference = (CtFieldReference) getReference();
            CtTypeReference<?> declaringType = ReflectionFactory.getInstance().clone(reference.getDeclaringType());
            declaringType.setActualTypeArguments(new ArrayList<>());
            return getFactory().wrap(declaringType);
        }
    }

    @Override
    public void tagDeclaringElement() {
        new DeclaringElementTagger(this).tagDeclaredBy();
    }
}

File: src/main/kotlin/org/codeontology/extraction/declaration/FieldEntity.kt
/*
Copyright 2021 Mattia Atzeni, Maurizio Atzori, Oliver Schmidtke
This file is part of CodeOntology.
CodeOntology is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
CodeOntology is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with CodeOntology.  If not, see <http://www.gnu.org/licenses/>
*/

package org.codeontology.extraction.declaration

import org.apache.jena.rdf.model.RDFNode
import org.codeontology.CodeOntology
import org.codeontology.Ontology
import org.codeontology.extraction.Entity
import org.codeontology.extraction.NamedElementEntity
import org.codeontology.extraction.ReflectionFactory
import org.codeontology.extraction.statement.FieldDeclaration
import org.codeontology.extraction.support.*
import spoon.reflect.declaration.CtField
import spoon.reflect.reference.CtFieldReference
import spoon.reflect.reference.CtTypeReference
import java.lang.Error

import java.lang.reflect.Field
import java.lang.reflect.GenericArrayType
import java.lang.reflect.Type
import java.lang.reflect.TypeVariable

class FieldEntity: NamedElementEntity<CtField<*>>, ModifiableEntity<CtField<*>>, MemberEntity<CtField<*>>, TypedElementEntity<CtField<*>> {
    constructor(field: CtField<*>): super(field)

    constructor(field: CtFieldReference<*>): super(field)

    override fun buildRelativeURI(): String {
        var uri = (reference?.simpleName ?: "")
        val declarator = getDeclaringElement()

        if(declarator != null) {
            uri = declarator.getRelativeURI() + SEPARATOR + uri
        }

        return uri
    }

    override fun getType(): RDFNode {
        return Ontology.FIELD_ENTITY
    }

    override fun extract() {
        tagName()
        tagLabel()
        tagType()
        tagDeclaringElement()
        tagJavaType()
        tagModifiers()
        if (isDeclarationAvailable()) {
            tagSourceCode()
            tagComment()
            tagAnnotations()
            if (CodeOntology.processStatements()) {
                tagDeclaration()
            }
        }
    }

    override fun getModifiers(): List<Modifier> {
        if (isDeclarationAvailable()) {
            return Modifier.asList(element?.modifiers ?: HashSet())
        }
        return try {
            Modifier.asList((reference as CtFieldReference<*>).modifiers)
        } catch (e: Exception) {
            ArrayList()
        } catch(e: Error) {
            ArrayList()
        }
    }

    override fun tagModifiers() {
        ModifiableTagger(this).tagModifiers()
    }

    override fun getJavaType(): TypeEntity<*> {
        var type: TypeEntity<*>?
        if (isDeclarationAvailable()) {
            type = getFactory().wrap(element!!.type)
        } else {
            type = getGenericType()
            if (type == null) {
                var typeReference: CtTypeReference<*>? = (reference as CtFieldReference<*>).type

                if(typeReference == null) {
                    typeReference = (reference as CtFieldReference<*>).declaringType
                }

                type = getFactory().wrap(typeReference)
            }
        }

        type!!.parent = getDeclaringElement()
        return type
    }

    private fun getGenericType(): TypeEntity<*>? {
        var result: TypeEntity<*>? = null
        if (isDeclarationAvailable()) {
            return null
        }
        try {
            val reference: CtFieldReference<*> = this.reference as CtFieldReference<*>
            val field: Field = reference.actualField as Field
            val genericType: Type = field.genericType

            if (genericType is GenericArrayType || genericType is TypeVariable<*>) {
                result = getFactory().wrap(genericType)
            }

        } catch (t: Throwable) {
            return null
        }

        return result
    }

    override fun tagJavaType() {
        JavaTypeTagger(this).tagJavaType()
    }

    override fun getDeclaringElement(): Entity<*>? {
        return if(isDeclarationAvailable()) {
            getFactory().wrap(element!!.declaringType)
        } else {
            val reference: CtFieldReference<*> = this.reference as CtFieldReference<*>
            val declaringType: CtTypeReference<*>? = ReflectionFactory.getInstance().clone(reference.declaringType)

            declaringType?.setActualTypeArguments<CtTypeReference<*>>(ArrayList())

            getFactory().wrap(declaringType)
        }
    }

    override fun tagDeclaringElement() {
        DeclaringElementTagger(this).tagDeclaredBy()
    }

    private fun tagDeclaration() {
        val declaration = FieldDeclaration(element!!)
        declaration.parent = this
        declaration.extract()
    }
}